library/minecraft_server_file.py

```diff
@@ -0,0 +1,354 @@
+    REQUIRED_ARGS = ['owner', 'group']
+
+    def __init__(self, module):
+        self.module = module
+        missing_args = [x for x in self.REQUIRED_ARGS
+                        if not self.module.params[x]]
+        if missing_args:
+            raise MissingArgsException(missing_args)
+
+        self.path = os.path.expanduser(module.params['path'])
+        self.owner = module.params['owner']
+        self.group = module.params['group']
+        self.mode = module.params['mode'] if module.params['mode'] else 0644
+
+    def file_stats_changed(self):
+        try:
+            st = os.stat(self.path)
+        except OSError:
+            # If the file doesn't exist or isn't read, this is the same as
+            # the stat values being wrong.
+            return True
+
+        current_mode = st.st_mode & stat.S_IMODE(st.st_mode)
+        # We cannot look up non-existant uids and gids when we compare the
+        # current state to desired state. Instead, get the usernames and
+        # groups now to make comparison easier.
+        current_uid, current_gid = self.module.user_and_group(self.path)
+        current_user = pwd.getpwuid(current_uid).pw_name
+        current_group = grp.getgrgid(current_gid).gr_name
+
+        return any([
+            self.owner != current_user,
+            self.group != current_group,
+            self.mode != current_mode,
+        ])
+
+    def correct_file_stats(self):
+        self.module.set_owner_if_different(self.path, self.owner, False)
+        self.module.set_group_if_different(self.path, self.group, False)
+        self.module.set_mode_if_different(self.path, self.mode, False)
+
+
+class ServerFile(object):
+    """
+    Abstract class representing any server file.
+    """
+    def __init__(self, module):
+        self.module = module
+        self.values = self.module.params['values']
+        self.stats = FileStats(module)
+
+    def content(self):
+        """
+        Return content in its native format (e.g. dict).
+        """
+        raise NotImplementedError
+
+    def content_str(self):
+        """
+        Return content as a string, for saving to a file.
+        """
+        return self.content()
+
+    def content_changed(self):
+        """
+        Indicates whether the content changed or not.
+        """
+        raise NotImplementedError
+
+
+class ServerProperties(ServerFile):
+    """
+    The Minecraft server.properties file.
+
+    The module "values" argument should be a list of strings in the format
+    "<name>=<value>".
+    """
+    def __init__(self, module):
+        super(ServerProperties, self).__init__(module)
+        self._content_changed = False
+        self.newlines = []
+        properties = {}
+
+        for a in self.values:
+            name, eq, val = a.partition('=')
+            if not eq:
+                raise BadValueException(name)
+            properties[name] = val
+
+        filein = file(self.stats.path)
+
+        for line in filein:
```

Comment:
This parser is somewhat brittle. Java properties files support more complex notations than `key=value`.

The [INI file lookup plugin](https://docs.ansible.com/ansible/playbooks_lookups.html#the-ini-file-lookup) offloads parsing to `configparser`:

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/lookup/ini.py#L57-L64

That `StringIO` hack is pretty dirty, but `configparser` will make it easier to compare changed values as well.